cmd/snap-update-ns/change.go

```diff
@@ -591,106 +586,124 @@ func neededChangesOld(currentProfile, desiredProfile *osutil.MountProfile) []*Ch
-			// Consume all other syn ethic entries without emitting either a
-			// mount, unmount or keep change.  This relies on the fact that all
-			// synthetic mounts are created by a mimic underneath a tmpfs that
-			// is detached, as coded above.
-			continue
-		case entry.OptBool("rbind") || entry.Type == "tmpfs":
-			// Recursive bind mounts and non-mimic tmpfs mounts need to be
-			// detached because they can contain other mount points that can
-			// otherwise propagate in a self-conflicting way.
-			if !entry.XSnapdDetach() {
-				entry.Options = append(entry.Options, osutil.XSnapdDetach())
+	if len(affectedTargetCreationDirs) != 0 {
+		entriesForMimicDir := map[string][]osutil.MountEntry{}
+		for parentTargetDir, entriesNeedingDir := range affectedTargetCreationDirs {
+			// first check if any of the mount entries for the changes will require
+			// creating a mimic
+			// we do this by checking if the file / dir that is to be mounted exists
+			// already in the form we need to to bind mount on top of it, if it
+			// doesn't then we need to create a mimic and so we then go looking for
+			// where to create the mimic
+			for _, entry := range entriesNeedingDir {
+				switch entry.XSnapdKind() {
+				case "":
+					// implicit dir creation, if the dir doesn't exist then we need
+					// a mimic
+					if !osutilIsDirectory(entry.Dir) {
+						neededMimicDir := findFirstRootDirectoryThatExists(parentTargetDir)
+						entriesForMimicDir[neededMimicDir] = append(entriesForMimicDir[neededMimicDir], entry)
+					}
+				case "file":
+					// if the file doesn't exist then we need a mimic
+					if !osutil.FileExists(entry.Dir) {
+						neededMimicDir := findFirstRootDirectoryThatExists(parentTargetDir)
+						entriesForMimicDir[neededMimicDir] = append(entriesForMimicDir[neededMimicDir], entry)
+					}
+				case "symlink":
+					// if the symlink doesn't exist then we need a mimic
+					if !osutil.IsSymlink(entry.Dir) {
+						neededMimicDir := findFirstRootDirectoryThatExists(parentTargetDir)
+						entriesForMimicDir[neededMimicDir] = append(entriesForMimicDir[neededMimicDir], entry)
+					}
+				}
 			}
-		case entry.OptBool("bind") && entry.XSnapdKind() == "file":
-			// Bind mounted files are detached. If a bind mounted file open or
-			// mapped into a process as a library, then attempting to unmount
-			// it will result in EBUSY.
-			//
-			// This can happen when a snap has a service, for example one using
-			// a library mounted via a bind mount and an absent content
-			// connection. Subsequent connection of the content connection will
-			// trigger re-population of the mount namespace, which will start
-			// by tearing down the existing file bind-mount. To prevent this,
-			// detach the mount instead.
-			if !entry.XSnapdDetach() {
-				entry.Options = append(entry.Options, osutil.XSnapdDetach())
+		}
+
+		// sort the mimic creation dirs to get the correct ordering of mimics to
+		// create dirs in
```

Comment:
I still think as mardy suggested that adding something like "with parents created before children"  would make this comment clearer

Reply:
Added.